### PR TITLE
Improved how Wasp prints output of underlying processes.

### LINKED ISSUE #793: Missing prefix (server/web app/db) for underlying process in `wasp start` when interrupted by other output
If we have following job output coming from server:
```
"Hello\n"
"hi\n"
"wohoo\n"
```

and there is a bit of pause between "hi\n" and "wohoo\n" and some other non-job output gets printed there, for example a warning by Wasp: "[wasp] watch out\n", then "wohoo" won't have "Server:" prefix in front of it, although we want it to have it.

So we will get:
```
Server: Hello
Server: hi
Server:
[wasp] watch out
wohoo
Server:
```

instead of 
```
Server: Hello
Server: hi
Server:
[wasp] watch out
Server: wohoo
Server:
```

This is due to how our logic in Wasp.Generator.Job.IO works, in `printPrefixedJobMessage` -> while we do account for different jobs interrupting each other, so for example if WebApp interrupts Server that will be ok, we don't account for some other input interrupting us.

How can we fix this? Some ideas:
1. When a message ends with "\n", don't print it immediately like we do now, instead "save" it for later, by remembering we have it waiting, and when the next messages comes from that same job, do smth like `print $ "\n" ++ prefix ++ message`. This idea is actually pretty neat, I think I like it.
2. Different idea is to recognize that Wasp has written warnings or some output and based on that add a prefix, same like we do output from other jobs. I like the first idea better though, it is simpler.

To elaborate first idea a bit, I imagine it working smth like this:
```hs
printPrefixedJobMessage :: RemainingNewlines -> Maybe J.JobMessage -> J.JobMessage -> IO RemainingNewlines
printPrefixedJobMessage remainingNewlines maybePrevJobMessage jobMessage = do
  let outHandle = getJobMessageOutHandle jobMessage
      prefix = makeJobMessagePrefix jobMessage
      content = getJobMessageContent jobMessage
      contentEndsWithNewline = "\n" `isSuffixOf` content
      content' = if contentEndsWithNewline then init content else content
      remainingNewlines' = if contentEndsWithNewline
                             then remainingNewlines `set` (J._jobType jobMessage)
                             else remainingNewlines `unset` (J._jobType jobMessage)

  let maybeAddPrefixAtStart =
        if (remainingNewlines `contains` J._jobType jobMessage)
          then "\n" <> prefix
          else ""
      addPrefixAfterSubstr substr = T.intercalate (substr <> prefix) . T.splitOn substr
      addPrefix = maybeAddPrefixAtStart . addPrefixAfterSubstr "\n" . addPrefixAfterSubstr "\r"

  T.IO.hPutStr outHandle $ addPrefix content'
  hFlush outHandle
  return remainingNewlines
```

Yup, I think this should mostly work!
Keep in mind this is a bit pseudocodish and I made up some functions and stuff here.
I imagine RemainingNewlines would be some kind of Set that contains job types -> if it has that job type in it, then there is a remaining newline from that type of job, otherwise there isn't.

### PR DESCRIPTION
**Pull request by Martinsos**
Fixes #793

### CODE REVIEW

File: waspc/cli/src/Wasp/Cli/Command.hs [hunk 1]
```diff
@@ -17,8 +17,7 @@ newtype Command a = Command {_runCommand :: ExceptT CommandError IO a}
 
 runCommand :: Command a -> IO ()
 runCommand cmd = do
-  errorOrResult <- runExceptT $ _runCommand cmd
-  case errorOrResult of
+  runExceptT (_runCommand cmd) >>= \case
```
Comment (Martinsos) [inline waspc/cli/src/Wasp/Cli/Command.hs:20]: Unrelated refactoring, but it is nice so I did it!

It is cool, with `\case` (which comes from LambdaCase extension) we can directly feed the result of monadic action into the case instead of having to extract it into a variable first, nice!
@matijaSos this is what I mentioned today to you.

File: waspc/src/Wasp/Generator/Job/Common.hs [hunk 2]
```diff
@@ -0,0 +1,24 @@
+module Wasp.Generator.Job.Common
```
Comment (Martinsos) [inline waspc/src/Wasp/Generator/Job/Common.hs:1]: I just moved this stuff from another file here, this is not new.

File: waspc/src/Wasp/Generator/NpmInstall.hs [hunk 3]
```diff
@@ -101,25 +102,27 @@ loadInstalledFullStackNpmDependencies dstDir = do
...
             J.JobOutput {} ->
-              printPrefixedJobMessage prevJobMsg jobMsg
-                >> go (Just jobMsg) (isWebAppDone, isServerDone) chan
+              printPrefixedJobMessage jobMsg
+                >> go (isWebAppDone, isServerDone) chan
             J.JobExit {} -> case J._jobType jobMsg of
-              J.WebApp -> go (Just jobMsg) (True, isServerDone) chan
-              J.Server -> go (Just jobMsg) (isWebAppDone, True) chan
+              J.WebApp -> go (True, isServerDone) chan
+              J.Server -> go (isWebAppDone, True) chan
```
Comment (Martinsos) [inline waspc/src/Wasp/Generator/NpmInstall.hs:125]: Introducing state monad allowed us to hide explicit state from here!
So now here we can call prefixed writing without having to know about how it keeps its state! It just works. Before state was "previous message", now it is not visible. Super cool!

File: waspc/src/Wasp/Generator/Job/Common.hs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    getJobMessageContent,
+  )
+where
+
+import qualified Data.Text as T
+import System.Exit (ExitCode (..))
+import System.IO (Handle, stderr, stdout)
+import qualified Wasp.Generator.Job as J
+
+getJobMessageOutHandle :: J.JobMessage -> Handle
```
[the remark this answers is not in the document]
Comment (shayneczyzewski) [reply waspc/src/Wasp/Generator/Job/Common.hs:12]: Tricky call. I think both are valid. I would say here slightly since they are more specific utility than just operating on the types themselves as generic helpers.

File: waspc/src/Wasp/Generator/Job/IO/PrefixedWriter.hs [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+  liftIO $ do
+    T.IO.hPutStr outHandle $ addPrefixWhereNeeded content'
+    hFlush outHandle
+  where
+    addPrefixWhereNeeded :: T.Text -> T.Text
+    addPrefixWhereNeeded =
+      let prefix = makeJobMessagePrefix jobMessage
+          addPrefixAfterSubstr substr = T.intercalate (substr <> prefix) . T.splitOn substr
+       in addPrefixAfterSubstr "\r" . addPrefixAfterSubstr "\n"
```
Comment (shayneczyzewski) [inline waspc/src/Wasp/Generator/Job/IO/PrefixedWriter.hs:74]: Why again are we splitting the `\r` and `\n` cases like this again, vs treating them as `\r\n`?

File: waspc/src/Wasp/Generator/Job/IO/PrefixedWriter.hs [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+      let addPendingNewlineToStartIfAny =
+            if jobType `S.member` jobsWithPendingNewline then ("\n" <>) else id
+          removeTrailingNewlineIfAny =
+            if contentEndsWithNewline then T.init else id
+       in addPendingNewlineToStartIfAny $ removeTrailingNewlineIfAny content
+    jobsWithPendingNewlines' =
+      if contentEndsWithNewline
+        then S.insert jobType jobsWithPendingNewline
+        else S.delete jobType jobsWithPendingNewline
+    contentEndsWithNewline = "\n" `T.isSuffixOf` content
```
Comment (shayneczyzewski) [inline waspc/src/Wasp/Generator/Job/IO/PrefixedWriter.hs:116]: I found some of the inner "if-ing" a bit hard to grok... is this any easier perhaps?

```haskell
applyPendingNewline ::
  JobsWithPendingNewline -> (J.JobType, T.Text) -> (JobsWithPendingNewline, T.Text)
applyPendingNewline jobsWithPendingNewline (jobType, content) =
  let content' = addPendingNewlineToStartIfAny $ removeTrailingNewlineIfAny content
      jobsWithPendingNewlines' = updateOp jobType jobsWithPendingNewline
   in (jobsWithPendingNewlines', content')
  where
    updateOp :: JobType -> Set JobType -> Set JobType
    updateOp = if contentEndsWithNewline then S.insert else S.delete

    addPendingNewlineToStartIfAny = if jobType `S.member` jobsWithPendingNewline then ("\n" <>) else id
    removeTrailingNewlineIfAny = if contentEndsWithNewline then T.init else id
    contentEndsWithNewline = "\n" `T.isSuffixOf` content
```

File: waspc/src/Wasp/Generator/NpmInstall.hs (hunk 3 above)
Comment (shayneczyzewski) [reply waspc/src/Wasp/Generator/NpmInstall.hs:125]: That is very cool you abstracted away the previous message - nice! 👍🏻

**Review by shayneczyzewski — COMMENTED** [review]
Hey @Martinsos - this is super cool! It made my head hurt, so I am taking a pause for now and will re-review tomorrow with fresh eyes/brain :D Dang Monad Transformers and all that stuff haha

One quick question, is there a clean way to write tests for this? I just did some playing in the repl, and may have found an undesired case, but I wonder if we can do some sort of fuzzing/prop testing (or even explicit unit tests) around this? Maybe not worth it?

Anyways, when I try the following:
```haskell
:set +m

import Wasp.Generator.Job.IO.PrefixedWriter (PrefixedWriter, printPrefixedJobMessage, runPrefixedWriter)
import qualified Wasp.Generator.Job as J

runPrefixedWriter $ do
  printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "line 1a" J.Stdout, J._jobType = J.Server }
  printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "one\n" J.Stdout, J._jobType = J.Db }
  printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "line 1b\n" J.Stdout, J._jobType = J.Server }
  printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "line 2\n" J.Stdout, J._jobType = J.Server }
```

I get this interleaving:
```bash
Server: line 1a
Db: oneline 1b
Server: line 2
```

I am guessing this second output line is not a case we want, or are we explicitly not protecting against that? Anyways, great work, I will resume tomorrow after the brain break!

File: waspc/src/Wasp/Generator/Job/IO/PrefixedWriter.hs (hunk 5 above)
Comment (Martinsos) [reply waspc/src/Wasp/Generator/Job/IO/PrefixedWriter.hs:74]: Pfuf, those are good questions!
`\r` is valid because it can be used to "reset the cartridge" -> so it moves cursor to the start of the line removing previous output and writes what comes after it. It can be used for stuff like showing a progress bar, or percentages growing. Prisma uses it in some of its output, I think npm also possibly (e.g. when installing stuff).

As for windows -> I know they use \r\n in the files, but maybe they don't in the shell?
I would maybe just ignore this one for now, and we can handle that when the windows come :D.

**Comment by Martinsos** [comment]
> Hey @Martinsos - this is super cool! It made my head hurt, so I am taking a pause for now and will re-review tomorrow with fresh eyes/brain :D Dang Monad Transformers and all that stuff haha
> 
> One quick question, is there a clean way to write tests for this? I just did some playing in the repl, and may have found an undesired case, but I wonder if we can do some sort of fuzzing/prop testing (or even explicit unit tests) around this? Maybe not worth it?
> 
> Anyways, when I try the following:
> 
> ```haskell
> :set +m
> 
> import Wasp.Generator.Job.IO.PrefixedWriter (PrefixedWriter, printPrefixedJobMessage, runPrefixedWriter)
> import qualified Wasp.Generator.Job as J
> 
> runPrefixedWriter $ do
>   printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "line 1a" J.Stdout, J._jobType = J.Server }
>   printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "one\n" J.Stdout, J._jobType = J.Db }
>   printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "line 1b\n" J.Stdout, J._jobType = J.Server }
>   printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "line 2\n" J.Stdout, J._jobType = J.Server }
> ```
> 
> I get this interleaving:
> 
> ```shell
> Server: line 1a
> Db: oneline 1b
> Server: line 2
> ```
> 
> I am guessing this second output line is not a case we want, or are we explicitly not protecting against that? Anyways, great work, I will resume tomorrow after the brain break!

You are right, this is problematic case! I on purpose gave up on handling it for now, but now that you write about it, I realized we did have that solved before, since we had information about the previous message. We could add that information back to the state of the PrefixedWriter and again handle it.
The thing is, we can't really defend ourselves completely -> there can be output from non-job, and we will have no idea about it, and we will get this issue, we can't avoid it.

Good thing though is, that in practice, messages usually don't came broken into pieces like that, mostly they come with newline at the end. Even if they do come broken in the middle, which is pretty / super rare from what I saw, they will then very likely come very quickly after each other, again making the chance of this happening very low. So I would say in practice, this is not likely to happen at all.

I think it still might be worth tracking the last message and defending against this use case, I will add that.

File: waspc/src/Wasp/Generator/Job/IO/PrefixedWriter.hs (hunk 6 above)
Comment (Martinsos) [reply waspc/src/Wasp/Generator/Job/IO/PrefixedWriter.hs:116]: I agree this part is a bit complex, I tried simplifying it couple of times but never got to perfect solution. This looks interesting I will try using it, maybe doing some hybrid!

File: waspc/src/Wasp/Generator/Job/IO/PrefixedWriter.hs (hunk 5 above)
Comment (shayneczyzewski) [reply waspc/src/Wasp/Generator/Job/IO/PrefixedWriter.hs:74]: Ah, yes, silly Prisma-like executables. Forgot they do carriage return stuff and some other funky ANSI symbols. Ok, makes sense! Yes, let's punt on Windows for now! :D

**Comment by shayneczyzewski** [comment]
> > Hey @Martinsos - this is super cool! It made my head hurt, so I am taking a pause for now and will re-review tomorrow with fresh eyes/brain :D Dang Monad Transformers and all that stuff haha
> > One quick question, is there a clean way to write tests for this? I just did some playing in the repl, and may have found an undesired case, but I wonder if we can do some sort of fuzzing/prop testing (or even explicit unit tests) around this? Maybe not worth it?
> > Anyways, when I try the following:
> > ```haskell
> > :set +m
> > 
> > import Wasp.Generator.Job.IO.PrefixedWriter (PrefixedWriter, printPrefixedJobMessage, runPrefixedWriter)
> > import qualified Wasp.Generator.Job as J
> > 
> > runPrefixedWriter $ do
> >   printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "line 1a" J.Stdout, J._jobType = J.Server }
> >   printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "one\n" J.Stdout, J._jobType = J.Db }
> >   printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "line 1b\n" J.Stdout, J._jobType = J.Server }
> >   printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "line 2\n" J.Stdout, J._jobType = J.Server }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I get this interleaving:
> > ```shell
> > Server: line 1a
> > Db: oneline 1b
> > Server: line 2
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I am guessing this second output line is not a case we want, or are we explicitly not protecting against that? Anyways, great work, I will resume tomorrow after the brain break!
> 
> You are right, this is problematic case! I on purpose gave up on handling it for now, but now that you write about it, I realized we did have that solved before, since we had information about the previous message. We could add that information back to the state of the PrefixedWriter and again handle it. The thing is, we can't really defend ourselves completely -> there can be output from non-job, and we will have no idea about it, and we will get this issue, we can't avoid it.
> 
> Good thing though is, that in practice, messages usually don't came broken into pieces like that, mostly they come with newline at the end. Even if they do come broken in the middle, which is pretty / super rare from what I saw, they will then very likely come very quickly after each other, again making the chance of this happening very low. So I would say in practice, this is not likely to happen at all.
> 
> I think it still might be worth tracking the last message and defending against this use case, I will add that.

I think you are right, it feels like a very unlikely scenario. I don't think we should invest much time on it now (possibly ever), but whether we want to create an issue or not is your call. I'm good either way, just wanted to note it and confirm! 👍🏻 This is really looking nice!

File: waspc/src/Wasp/Generator/Job/IO/PrefixedWriter.hs (hunk 5 above)
Comment (Martinsos) [reply waspc/src/Wasp/Generator/Job/IO/PrefixedWriter.hs:74]: Hey `\r` is not that weird, we are also going to use it for `npm install` progress indicator probably :D!
Btw I am going to leave a TODO here to think more about Windows in the future.

File: waspc/src/Wasp/Generator/Job/IO/PrefixedWriter.hs (hunk 6 above)
Comment (Martinsos) [reply waspc/src/Wasp/Generator/Job/IO/PrefixedWriter.hs:116]: I went with something in between and ended up with this:
```hs
applyPendingNewline ::
  JobsWithPendingNewline -> (J.JobType, T.Text) -> (JobsWithPendingNewline, T.Text)
applyPendingNewline jobsWithPendingNewline (jobType, content) = (jobsWithPendingNewline', content')
  where
    content' = addPendingNewlineToStartIfAny $ removeTrailingNewlineIfAny content
      where
        removeTrailingNewlineIfAny =
          if contentEndsWithNewline then T.init else id
        addPendingNewlineToStartIfAny =
          if jobType `S.member` jobsWithPendingNewline then ("\n" <>) else id
    jobsWithPendingNewline' = updateOp jobType jobsWithPendingNewline
      where
        updateOp = if contentEndsWithNewline then S.insert else S.delete
    contentEndsWithNewline = "\n" `T.isSuffixOf` content
```

What does it look like?

**Comment by Martinsos** [comment]
> > > Hey @Martinsos - this is super cool! It made my head hurt, so I am taking a pause for now and will re-review tomorrow with fresh eyes/brain :D Dang Monad Transformers and all that stuff haha
> > > One quick question, is there a clean way to write tests for this? I just did some playing in the repl, and may have found an undesired case, but I wonder if we can do some sort of fuzzing/prop testing (or even explicit unit tests) around this? Maybe not worth it?
> > > Anyways, when I try the following:
> > > ```haskell
> > > :set +m
> > > 
> > > import Wasp.Generator.Job.IO.PrefixedWriter (PrefixedWriter, printPrefixedJobMessage, runPrefixedWriter)
> > > import qualified Wasp.Generator.Job as J
> > > 
> > > runPrefixedWriter $ do
> > >   printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "line 1a" J.Stdout, J._jobType = J.Server }
> > >   printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "one\n" J.Stdout, J._jobType = J.Db }
> > >   printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "line 1b\n" J.Stdout, J._jobType = J.Server }
> > >   printPrefixedJobMessage $ J.JobMessage { J._data = J.JobOutput "line 2\n" J.Stdout, J._jobType = J.Server }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I get this interleaving:
> > > ```shell
> > > Server: line 1a
> > > Db: oneline 1b
> > > Server: line 2
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I am guessing this second output line is not a case we want, or are we explicitly not protecting against that? Anyways, great work, I will resume tomorrow after the brain break!
> > 
> > 
> > You are right, this is problematic case! I on purpose gave up on handling it for now, but now that you write about it, I realized we did have that solved before, since we had information about the previous message. We could add that information back to the state of the PrefixedWriter and again handle it. The thing is, we can't really defend ourselves completely -> there can be output from non-job, and we will have no idea about it, and we will get this issue, we can't avoid it.
> > Good thing though is, that in practice, messages usually don't came broken into pieces like that, mostly they come with newline at the end. Even if they do come broken in the middle, which is pretty / super rare from what I saw, they will then very likely come very quickly after each other, again making the chance of this happening very low. So I would say in practice, this is not likely to happen at all.
> > I think it still might be worth tracking the last message and defending against this use case, I will add that.
> 
> I think you are right, it feels like a very unlikely scenario. I don't think we should invest much time on it now (possibly ever), but whether we want to create an issue or not is your call. I'm good either way, just wanted to note it and confirm! 👍🏻 This is really looking nice!

@shayneczyzewski I also added this now to the prefixing logic, and while doing it, I also found another case with a missing newline sometimes -> anyway, this should be covered well now! I think it is good we added it, I feel the output is now more robust. It was also easy to add additional tracking of last message, which was cool.

While at it I also improved the looks of output a bit: gave nice color to "DB" and colored "stderr" in yellow to be more visible.

Here is a screenshot of what it look like now:
image

**Comment by Martinsos** [comment]
Well I just realized from the screenshot above that I completely forgot that some messages can come from stderr while some can come from stdout! And those we also don't want to mix!

I now modified code to also take that into account. So now it doesn't care just about job type, but about something called "output".

Uff, I think this should be it! Maybe take a quick look at the `Output` type @shayneczyzewski and tell me if that seems sound.

File: waspc/src/Wasp/Generator/Job/IO/PrefixedWriter.hs (hunk 5 above)
Comment (shayneczyzewski) [reply waspc/src/Wasp/Generator/Job/IO/PrefixedWriter.hs]: Haha, yeah it's true `\r` is fine for interactive outputs, I was just joking since I spaced on it. :D TODO for Windows makes sense. 👍🏻

File: waspc/src/Wasp/Generator/Job/IO/PrefixedWriter.hs (hunk 6 above)
Comment (shayneczyzewski) [reply waspc/src/Wasp/Generator/Job/IO/PrefixedWriter.hs:160]: Looks good to me!

**Review by shayneczyzewski — APPROVED** [review]
Hey @Martinsos nice updates! Good catch on tracking `stdout` in the decision-making for prefixes. I think adding that bool to `Output` makes sense.

Glad to see tracking the last job message was made easier with `PrefixedWriterState` as well.

`printPrefixedJobMessage` does still feel a bit hard to follow to me sometimes, but I think you minimized as much complexity as you can and the detailed comment above helps readers as well. It is just a hard problem to solve, and the helpers do aid in understanding what is going on, so I think much of it is essential complexity.

I did a quick test and what is here seems to function nicely. 👍🏻 The only thing I would consider is possibly adding a space between the `JobType` and "(stderr)", so you would go from "Db(stderr):" to "Db (stderr):", but that may just be my visual preference.

**Comment by Martinsos** [comment]
@shayneczyzewski you are right that printPrefixedJobMessage is still a bit too complex.
I now refactored it a bit more, take a look, I believe it is a bit more readable.
I also changed the name to printJobMessagePrefixed, since that is more correct -> we are not printing already prefixed job message, we prefix it ourselves and then print it.

As for space between Server and (stderr) -> there was actually space there before, but I removed it because it felt a bit like it was part of the input to the right. At least I subjectively thought so :D. So maybe let's leave it like this for now and we can change in the future if we figure out it is not the best?

**Comment by shayneczyzewski** [comment]
> @shayneczyzewski you are right that printPrefixedJobMessage is still a bit too complex. I now refactored it a bit more, take a look, I believe it is a bit more readable. I also changed the name to printJobMessagePrefixed, since that is more correct -> we are not printing already prefixed job message, we prefix it ourselves and then print it.
> 
> I am sure it could be improved further, but it feels like I reached the threshold where refactoring pays off, I feel going further the ratio of invested time and what I got would not be the best.
> 
> As for space between Server and (stderr) -> there was actually space there before, but I removed it because it felt a bit like it was part of the input to the right. At least I subjectively thought so :D. So maybe let's leave it like this for now and we can change in the future if we figure out it is not the best?

Looking great! I like the refactor to simplify things a bit more. I think this will work just fine. 👍🏻 

No sweat on space, we can leave it as-is and see what others think. All good by me. 🚀
